src/index/DeltaTriples.cpp

```diff
@@ -0,0 +1,189 @@
+// Copyright 2023 - 2024, University of Freiburg
+//  Chair of Algorithms and Data Structures.
+//  Authors:
+//    2023 Hannah Bast <bast@cs.uni-freiburg.de>
+//    2024 Julian Mundhahs <mundhahj@tf.uni-freiburg.de>
+
+#include "index/DeltaTriples.h"
+
+#include "absl/strings/str_cat.h"
+#include "index/Index.h"
+#include "index/IndexImpl.h"
+#include "index/LocatedTriples.h"
+#include "parser/TurtleParser.h"
+
+// ____________________________________________________________________________
+LocatedTriples::iterator& DeltaTriples::LocatedTripleHandles::forPermutation(
+    Permutation::Enum permutation) {
+  switch (permutation) {
+    case Permutation::PSO:
+      return forPSO_;
+    case Permutation::POS:
+      return forPOS_;
+    case Permutation::SPO:
+      return forSPO_;
+    case Permutation::SOP:
+      return forSOP_;
+    case Permutation::OSP:
+      return forOSP_;
+    case Permutation::OPS:
+      return forOPS_;
+    default:
+      AD_FAIL();
+  }
+}
+
+// ____________________________________________________________________________
+void DeltaTriples::clear() {
+  triplesInserted_.clear();
+  triplesDeleted_.clear();
+  locatedTriplesPerBlockInPSO_.clear();
+  locatedTriplesPerBlockInPOS_.clear();
+  locatedTriplesPerBlockInSPO_.clear();
+  locatedTriplesPerBlockInSOP_.clear();
+  locatedTriplesPerBlockInOSP_.clear();
+  locatedTriplesPerBlockInOPS_.clear();
+}
+
+// ____________________________________________________________________________
+std::vector<DeltaTriples::LocatedTripleHandles>
+DeltaTriples::locateAndAddTriples(
+    ad_utility::SharedCancellationHandle cancellationHandle,
+    std::span<const IdTriple<0>> idTriples, bool shouldExist) {
+  ad_utility::HashMap<Permutation::Enum, std::vector<LocatedTriples::iterator>>
+      intermediateHandles;
+  for (auto permutation : Permutation::ALL) {
+    auto& perm = index_.getImpl().getPermutation(permutation);
+    auto locatedTriples = LocatedTriple::locateTriplesInPermutation(
+        // TODO<qup42>: replace with the method for update block metadata once
+        // integration is done
+        idTriples, perm.metaData().blockData(), perm.keyOrder(), shouldExist,
+        cancellationHandle);
```

Comment:
Isn't it true that the `blockData()` already lives inside the `LocatedTriple`, and the `keyOrder` should be a member of the `LocatedTriple`.
On the other hand I currently  have forgotten, which state for more than one triple is stored in the `LocatedTriples`at all...

Reply:
Ah, I think I know what I want to say:

The `LocatedTriples` should know about the above members, 
and then you immediately call a function `locatedAndAdd` on that object, which gives you back the handles and basically does the two steps that are performed here.

Reply:
I mostly don't understand what you want here.
- `getAugmenteMetadata` is already present, but doesn't work yet because it is not initialized. This will be part of the integration.
- Do I understand you correctly that you propose that every `LocatedTriple` should the Permutation to which it belongs?
- It is unclear to me how that would work. `LocatedTriplesPerBlock` manages all `LocatedTriple`s for a permutation. `DeltaTriples` does the management of `LocatedTriplesPerBlock` for all permutation and exposes an interface that is no longer concerned with individual permutations. For this it needs the handles for all the permutations of each triple. How exactly do you propose to move this into `LocatedTriples`?